### ISSUE TITLE
[Returned for Revisions] SDL 0304 - Start Video Stream With OnTouchEvent Permission

### ISSUE THREAD

**Issue by theresalech** [smartdevicelink/sdl_evolution#1014]
Hello SDL community,

The review of "SDL 0304 - Start Video Stream With OnTouchEvent Permission" begins now and runs through February 02, 2021. 

Previous reviews of this proposal took place on:
- May 20 - May 26, 2020 
- June 24 - July 21, 2020
- August 19 - September 1, 2020
- October 14, 2020 - October 20, 2020

The proposal is available here:

https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0304-stream-ontouchevent.md

Reviews are an important part of the SDL evolution process. All reviews should be sent to the associated Github issue at:

https://github.com/smartdevicelink/sdl_evolution/issues/1014

What goes into a review?

The goal of the review process is to improve the proposal under review through constructive criticism and, eventually, determine the direction of SDL. When writing your review, here are some questions you might want to answer in your review:

* Is the problem being addressed significant enough to warrant a change to SDL?
* Does this proposal fit well with the feel and direction of SDL?
* If you have used competitors with a similar feature, how do you feel that this proposal compares to those?
* How much effort did you put into your review? A glance, a quick reading, or an in-depth study?
Please state explicitly whether you believe that the proposal should be accepted into SDL.

More information about the SDL evolution process is available at

https://github.com/smartdevicelink/sdl_evolution/blob/master/process.md

Thank you,
Theresa Lech

Program Manager - Livio
theresa@livio.io

**Comment by joeljfischer** [comment]
1\. I don't see how this proposal could be accepted without corresponding Java Suite impact. We've done a lot of work to keep the libraries in alignment and we have lately not accepted changes only to one platform that clearly has impact for multiple platforms.

2\. The code to monitor the permissions and check if it's allowed is unnecessary due to the existence of the permission manager. The example should be refactored to use the permission manager instead.

**Comment by joeygrover** [comment]
3\. 

>Potentially this code can have an impact on existing head units that would allow navigation apps to start the video stream without touch permissions.

I think we also have to consider the case that the permission is never updated but is always defaulted to true. Because of this we need confirmation from all OEM members that this solution will not break their integrations.

4\. I'm a bit confused on what the library should do if the user moves the app to HMI_FULL but the permission is not yet granted; in this case the `OnPermissionsChange ` would never be sent. Would the UI just be a blank screen? Is there some sort of timeout that would occur if the permission is never granted?

5\. 
>The video streamable status notification should provide very similar functionality, however it's not available on all nav app-capable head units.

Should we then add a specific RPC version threshold to the usage of the permission requirement instead of adding this as an additional flag that should no longer be required since the inclusion of the streamable state?

6\. As an alternative, the library could retry to start the service after they receive the NAK right? Maybe it makes sense to investigate this approach in the case that OEMs integrations don't send the `OnPermissionsChange` as discussed in point 3.

**Comment by theresalech** [comment]
The Steering Committee voted to defer this proposal until the author is able to respond to the comments provided on the review issue.  Once the author is able to respond, he can email SDLC@smartdevicelink.com, and the proposal can be brought back under Steering Committee review.

**Comment by kshala-ford** [comment]
@joeygrover @joeljfischer 

1\. In fact this was the result from an internal meeting to submit this iOS modification to the open which I have done. I don't see any reason not to add the Java Suite. The revision would be to add Java Suite as an Impacted Platform. It should align with iOS identifying the OnTouchEvent permission state. Details are not finalized yet as per other items.

2\. That's fine. I can make that revision.

3\. I think I don't understand your question. After the app registration, SDL Core sends an `OnPermissionsChange` notification together with an `OnHMIState` notification. If this `OnTouchEvent` permission is granted (either by default or user choice from previous usage) then the first permission notification should contain this information. This RPC is not part of the base functional groups therefore if an OEM sends `OnPermissionsChange` only after app registration (because the policy table is static), the OEM must make this RPC enabled by default (at least for the specific app). Otherwise, with Policy table updates the permission change notification would inform the app once the RPC becomes available. 

4\. The point is not to even start the video service session hence no video data can be sent. In 3. I described what should happen if the app is known. If the registering app is not known it should not have permissions to `OnTouchEvent` and this app would never be able to start the video service; Core would return with `NAK`. This is a major issue using navigation apps for the first time or before the policy table is updated to recognize the app.

5\. According to SDL-0150 `videoStreamingState` is not policy controlled. It resolves issues with other application's HMI state. I reviewed SDL Core code and the only relevant code I could find was at `OnAllowSDLFunctionality` on device level. The major issue is that an app doesn't know when it has permission to start a video/audio service. `videoStreamingState` sounds like the much more obvious parameter to listen for if it would serve the logic we need. We may want to revisit the streaming state parameters and don't indicate `STREAMABLE` if the apps are actually not streamable. I don't know if that is a design issue / bug or requires a separate proposal or if it should be part of this proposal. I would like to include @yang1070 in this conversation as he's the author of the other proposal.

For Core <5.0 there's no other solution I could find to resolve the issue. I am fine to add an RPC version threshold though the problem still remains with Core >=5.0.

6\. Please note https://github.com/smartdevicelink/sdl_evolution/pull/1041. The PR is still WIP but we did some analysis for a retry mechanism. The StartService request can still fail even though the app has all the permissions and certificates for protected service. The app/library should not blindly start sending StartService requests after entering HMI_FULL. Instead it should wait until it reached highest confidentiality with the collectable information. Again `OnPermissionsChange` should always be sent at least once after app registration.

**Comment by Sohei-Suzuki-Nexty** [comment]
7.Could you provide a sequence diagram of how the proposal would improve the problem?

**Comment by joeljfischer** [comment]
1. 👍 
2. 👍

**Comment by Sohei-Suzuki-Nexty** [comment]
7.Supplement
>However, for nav apps that connect to a head unit for the first time, the app ID is unknown and therefore it's not clear if the app is allowed to video stream or to receive touch events. 

I want to clarify the state why the app ID is unknown here.
If you know the app ID, wouldn't the problem happen?

**Comment by joeygrover** [comment]
3\. I might be confused here, but to my understanding it is not required to enable any sort of permission modeling for SDL. In such a case that no permission modeling is in place all RPCs would be enabled for all applications. I don't believe it is mandated that Core must send an `OnPermissionChange` to a connected application informing them they have access to all RPCs. If Core doesn't send the `OnPermissionChange` because all permissions are granted by default, the permission check would always return false. However, as I said maybe my understanding is off here; if so can you share where this requirement is?

4\. I understand that the idea is not to stream anything form the mobile side until the permission is granted. However, in the case that a PTU didn't happen in time for the user to move the app to HMI_FULL the app would simply appear unresponsive to the user. Since the app can't start streaming because it doesn't have the `OnTouchEvent` permission granted, it seems it will silently fail while the IVI system would be waiting to receive video data. In Ford's case it would just have the dialog "Waiting for device.." until the user navigated away from that app and screen. 

If this requirement is used for future RPC versions, I think the HMI guidelines should be updated to reflect what would happen in this case to give a better UX instead of just a waiting type dialog that will never succeed.

5\. I personally think it makes much more sense to alter Core to only send `streamable="true" when the app is actually streamable. I understand your point that older RPC spec versions can't support this because it wasn't tied to such a requirement in the past. 

6\. Adding a way to retry starting a service would be a really nice feature to add to SDL in general. I don't want to debate another proposal inside this review so I would like to only comment on the idea of using a retry function in place of the permission check.

I think for a lot of potential use cases the touch permission makes sense that it must be granted, but I also don't think that holds true to the future where we might be able to stream content to different displays. So I want to make sure the solution we choose doesn't just solve the problem today, but it also doesn't chain us to something unworkable in the future.

Depending on what requirements we can gather for number 3, I think we have to still consider the retry strategy as a potential solution. Obviously if there are requirements that the `OnPermissionChange` is sent at least once, this weakens the argument to use a retry instead. If it is not a requirement, then adding a retry will not negatively affect any implementation while potentially solving this problem through allowing a PTU to happen at some point during the app's retry sequence.

**Comment by kshala-ford** [comment]
3\. It took me a while to go through all the content and find out the right answer. I don't think there is a specification describing your behavior. So I went through the Core and mobile implementations to understand the existing behavior.

In the smartDeviceLink.ini I found the setting `EnablePolicy` in `[Policy]` which seems to enable/disable SDL Core using a policy manager. I saw that Core handles disabled policy as a warning `LOG4CXX_WARN(logger_, "System is configured to work without policy functionality.");` (see https://github.com/smartdevicelink/sdl_core/blob/master/src/components/application_manager/src/application_manager_impl.cc#L2541). Core is also returning with `WARNING` when registering on a system with disabled policy. I don't think this is a recommend/supported configuration.

Without a policy manager, the policy handler does not trigger an `OnPermissionsChange` upon App registration. This causes confusion now on the mobile side. In the mobile libraries the permission managers rely on `OnPermissionChange` notifications. Without this RPC the manager treats all RPCs as not allowed.

```objc
- (BOOL)isRPCAllowed:(NSString *)rpcName {
    if (self.permissions[rpcName] == nil || self.currentHMILevel == nil) {
        return NO;
    }
```

```java 
private boolean isRPCAllowed(@NonNull FunctionID rpcName, Map<FunctionID, PermissionItem> permissionItems, HMILevel hmiLevel){
        PermissionItem permissionItem = permissionItems.get(rpcName);
        if (hmiLevel == null || permissionItem == null || permissionItem.getHMIPermissions() == null || permissionItem.getHMIPermissions().getAllowed() == null){
            return false;
```

So without any permissions notification the managers return `false`/`NO` for any RPC. As I agreed to use the permission manager, the libraries would need to be changed to reflect Core behavior.

We will need to decide if disabled policies should be supported by SDL. If we want to, we should change Core to return SUCCESS instead of WARNING and also modify the libraries to init with all RPCs permitted by default. Otherwise we may need to be more restrictive in Core prevent it from launching or Core should load a "dummy" policy manager which blindly allows everything and sends a proper OnPermissionsChange.

4\. Please take note that this proposal does not delay the video stream. It holds back the StartService(0xB) avoiding NAK until we can be confident the system will return ACK. This is obviously a precondition to actually send video data. 

4\.a. With 0xB being force protected you can't start a video service without valid app permission. Any attempt would result in StartServiceNAK. `OnTouchEvent` is permitted together with the reception of the crypto assets required by the security manager. Both come together in a PTU.

4\.b. If 0xB is unprotected the video service can start immediately. However, OnTouchEvent is required otherwise the user cannot perform touch input to the app. The app would simply be unresponsive. Therefore even in this case the app shouldn't start streaming.

5\. Yes I completely agree. I am happy to alter the streamable logic in the proposal. There may be alternatives to consider like extending `OnPermissionsChange` to also include permitted services. I'm happy about any feedback. As you say we will also need a solution for SDL systems on the road. I'm happy to focus on a mobile only solution for older SDL versions with this `OnTouchEvent` suggestion. With future SDL releases we should have a better solution providing more specific information.

6\. As already mentioned a retry mechanism can help for errors that cannot be identifed otherwise. It will add complication and must be tested against existing head units. Similar to sdl_ios retry delay we will need to confirm that existing head units can properly handle multiple subsequent StartService attempts.

**Comment by theresalech** [comment]
We did not have a quorum present during the 2020-07-07 Steering Committee meeting, so voting on this proposal could not take place.  This proposal will remain in review until 2020-07-14.

**Comment by Sohei-Suzuki-Nexty** [comment]
7.
@kshala-ford Thanks for replying
>Initially this database might not contain an mobile navigation app ID.

Does this mean that the RAI has completed successfully even if the appID is unknown?

**Comment by kshala-ford** [comment]
@Sohei-Suzuki-Nexty Yes this is possible with SDL policy system if the IVI relies on a policy backend such as SDL Server. See below diagram which is pretty long but covers three different scenarios 

1. the ID is not known and the policy update finishes before the user starts the app or 
2. the ID is not known and the policy update did not finish when the user starts the app and 
3. the ID is already known.

![image](https://user-images.githubusercontent.com/9195582/87298856-39028500-c50b-11ea-908e-5371d6753f14.png)

**Comment by joeygrover** [comment]
3\. I think this was an oversight in the design of the PermissionManager. Joel and I have talked about this offline a little bit including your comments and we think there is a path forward. The following should not be tied to this proposal directly but to move the process along I think it's important that we at least have a potential plan of action.

If an `onPersmissionChange` hasn't been received at all, the manager should simply return true that the requested RPCs are ok to use. Obviously if they truly weren't the IVI would be able to handle that and return the `DISALLOWED` status code. We would also need to add a new flag and retrieval method that would keep track of if the manager has received an `onPersmissionChange` yet. This could either be a flag or a new state, but thinking of it has initialized or not. Then it could be checked first to know if the permission manager will respond with anything of certainty or simply a value that won't break implementations.  

We can then add a new flag in the RAI response that will state if the module has policy control turned on or not. Using RPC spec version and this new flag we will be able to better determine the validity of the `PermissionManager`.

4\. I understand that the `StartService` is potentially "held back" or delayed, which in turn delays anything appearing on screen because no video data is sent. This is a potential issue still. I understand that a premature NAK is seen as worse and don't disagree, but I would like better acknowledgement in the proposal that there is potential for a longer delay because of the PTU requirement.  OEMs will need to understand what they should in this case. Put up a dialog? Show a blank screen? How long should such a thing be on screen before the HMI times out?

4b\. While in practice I agree, I'm not sure we have requirements that state video streaming must be accompanied by touch enabled. Theoretically an HMI could be implemented to be all voice control or through different buttons that are not related to touch points on the video stream. So because this proposal is explicitly relying on the requirement to have touch permissions enabled it does change the scope of what implementations can/must do. 


5\.  👍 


6\. Yea I think of the retry as a catch all solution. Obviously that doesn't mean it's the best solution for all problems, but does solve them. So the question was really, do we find a retry a "good enough" solution or not for this case. The more I think about the proposed solution the more I see this is going to be a complicated situation that has many different corner case solutions. If we take the item in discussion point 3 and 5 with this point, we have a lot of different logic put in place to try to get us to a fixed state. For the sake of this proposal, I'm fine with simply agreeing that we need to investigate and hopefully add a retry feature through a new proposal. 

7\. I want to add that in the scenario being referenced, Core and the HMI will know the App ID at the time of receiving the `RegisterAppInterface` request. What it doesn't know is linking that App ID to any sort of policy entry. So Core and HMI have the App ID when receiving the RAI request, but they don't know what to do with it.

**Comment by E-SAITO-TMC** [comment]
Is there anything I should be aware of if this proposal is whitelisted?

**Comment by theresalech** [comment]
The Steering Committee voted to keep this proposal in review to allow for additional conversation on the review issue.

**Comment by theresalech** [comment]
This proposal has been deferred in accordance with [this section](https://github.com/smartdevicelink/sdl_evolution/blob/master/process.md#for-consideration) of the SDL Evolution Process document.  The proposal will remain deferred until the author has acknowledged (via email to sdlc@smartdevicelink.com) that they have time to respond to comments in the next week.

**Comment by jordynmackool** [comment]
The author has advised that he is able to respond to the comments provided on the issue. This proposal is ready to be brought back into review, pending the Steering Committee vote.

**Comment by theresalech** [comment]
The Steering Committee voted to bring this proposal back into review, as the author is able to respond to the latest comments. This proposal will remain in review until August 25, 2020.

**Comment by kshala-ford** [comment]
3\. In the case of app resumption, your suggestion adds a small time frame where `OnTouchEvent` would be permitted at Ford systems. At app resumption the RPCs are sent as followed:
1. `RegisterAppInterface` response
2. `OnHMIStatus` (with `HMI_FULL`)
    - _During this time OnTouchEvent is permitted and therefore breaks with this proposal_
3. `OnPermissionsChange` notification

I want to ask again if the SDLC Members want to officially support non-policy controlled SDL Core. Again, Core returns a WARNING on registration if policy is disabled.

4\. 

> I understand that the StartService is potentially "held back" or delayed, which in turn delays anything appearing on screen because no video data is sent.

I repeat from my previous comment: Please take note that this proposal does not delay the video stream.  The potential issue you describe is not valid. Furthermore starting streaming before touch event is available causes issues where touch input is not applied to the app UI because the app can't receive OnTouchEvent notifications.

4b\. In SHAID the navigation apps obviously desire touch permissions. You say we don't have requirements that make video streaming require touch events. However we don't have any opposite requirements. This is clearly a lack in the SDL logic and requirements. Fact is that we have touch screens for video streaming. Your assumption would be valid if SDLC members have no vehicle with touch screen but desire navigation apps. These members could remove OnTouchEvent permissions for the nav app ID in their SDL server instance. The moment they have one vehicle with touch screen they would need to add this permission back again. 

I don't think anyone uses SDL in this way but I am okay adding barriers like Ford only or SDL version to avoid any hypothetical problem. Again the lack of information and official support is the actual problem. We should resolve this when executing the agreement in 5.

6\. I must be honest I don't see this to be complicated if the condition of starting the video stream is made transparent e.g. as a separate method in the streaming manager. A retry mechanism would be a fallback for basically any case but definitely adds delay in launching and adds stress to head units potentially causing race conditions when trying to start the stream over and over again before/while/after PTU. Also Ford will need to test production versions if they are actually capable of a retry mechanism when including the pre-PTU phase.

7\. Thanks for clarifying. With "App ID (not) known" I mean SDL has (no) policies for the given app ID.

**Comment by theresalech** [comment]
The Steering Committee voted to keep this proposal in review until our next meeting, to allow for additional time for discussion on the review issue.

**Comment by joeygrover** [comment]
3\. I think this relates to the conversation we had in the workshop for the Audio I/O Manager, there is nothing stating Core must have an accompanied policy server implementation. Some companies use other ways to protect their system from apps including whitelisting only specific applications through a non SDL means. So this would be an explicit requirement we would have to make but I'm fairly certain there are production units on the road today that do not support policies. I think we really need other OEMs to weigh in on this point because if we all agree that we should enforce policy table permissions then this would no longer be an issue.


4\. I'm sorry but it is a valid situation and I think there is a misunderstanding. For example, policy table has not been updated to include touch permissions for specific app that has just registered, IVI system which is in production toady has user select application, the application moves to HMI_FULL and thus the HMI displays _some screen_, the app has not received the touch permission yet so it waits until it does to send the `StartService` for video streaming, however the IVI system doesn't know this so it sits and displays either a blank screen or some sort of dialog. The `StartService` is not sent until the `OnTouchEven` permission is received, or in other words, the video service is potentially delayed until it receives the correct permissions.

4b\. This is again not an explicit requirement which is why I brought it up, not that I don't agree it should be one. The proposal should simply call this out as a defined requirement in order to support the inclusion of such a change in the mobile libraries.

**Comment by kshala-ford** [comment]
3\. It'll be impossible for me to attend and request feedback from the other OEMs but having a clear agreement whether mobile navigation apps always require touch capabilities or not. If that helps I would like to add a suggestion of how the app can handle user input. Adding a flag into the streaming builder / configuration that is called `requiresTouchInput` could help allowing apps to describe support or not. 

I can for sure say the navigation applications Ford partners with would not work without touch input. This includes the touch screen or anything that could be translated to it (touch panel/pad or knob). Basically this means "Can the app handle user input based on content shown in the video stream?"

4\. You're missing a very important piece in this puzzle which is "at what moment does the system accept starting the video service? In my situation video service is listed in `ForceProtectedService` therefore the video service can't be started at the time you describe thus the HMI must implement. I think you're describing a possible delay that would be possible in non-protected video service but with external policy system. I don't think you talk about Ford IVI in production because the sequence would be wrong. Let me try to explain it differently in a time table assuming the application was never registered before on the system and policy table update would be required (if external policy is used):

<table>
    <tbody><tr>
     <td>ForceProtectedService</td>
     <td colspan="2">Yes</td>
     <td colspan="2">Yes</td>
     <td colspan="2">No</td>
     <td colspan="2">No</td>
    </tr>
    <tr>
     <td>External/Local policy</td>
     <td colspan="2">External</td>
     <td colspan="2">Local</td>
     <td colspan="2">External</td>
     <td colspan="2">Local</td>
    </tr>
    <tr>
     <td>Permission</td>
     <td>StartService</td>
     <td>TouchEvent</td>
     <td>StartService</td>
     <td>TouchEvent</td>
     <td>StartService</td>
     <td>TouchEvent</td>
     <td>StartService</td>
     <td>TouchEvent</td>
    </tr>
    <tr>
     <td>1. App registers on the system</td>
     <td>NAK</td>
     <td>No</td>
     <td>NAK</td>
     <td>No</td>
     <td>NAK</td>
     <td>No</td>
     <td>NAK</td>
     <td>No</td>
    </tr>
    <tr>
     <td>2. App is user selected &gt; HMI_FULL</td>
     <td>NAK</td>
     <td>No</td>
     <td>ACK</td>
     <td>Yes</td>
     <td>ACK</td>
     <td>No</td>
     <td>ACK</td>
     <td>Yes</td>
    </tr>
    <tr>
     <td>3. System gets policy update</td>
     <td>ACK</td>
     <td>Yes</td>
     <td>ACK</td>
     <td>Yes</td>
     <td>ACK</td>
     <td>Yes</td>
     <td>ACK</td>
     <td>Yes</td>
    </tr>
   </tbody></table>

My concerns which I try to address are related to the first column with force protected service and external policies. A navigation app would not have the chance to start the video stream when transitioning into HMI_FULL.

You describe the scenario in the third column **without** protected video service **but** with external policies. In that case the app could start the video stream but will likely lack touch event permissions because this RPC is not part of the base policies in your functional groups.

Ford did resolve the problem by adding the popup "Waiting for Device" until the app started the video service and pushed the first frames out. This behavior is in production already. I agree in the other case the IVI might not have this alternative UI because it was never necessary. The behavior is undefined and could be a black/green screen depending on the video decoder (if already visible).

Before we run in circles I would like to suggest to add following conditions:

a. did the library find a security manager for the IVI make?
b. does the app require touch input?
c. is OnTouchEvent not allowed?

The suggestion is to keep the library wait starting the video stream until all three conditions are met. However, if there is no security manager it is not necessary to hold back the service request because it'll likely be accepted. Also it should not be held back if the app doesn't care about touch input.

4b. Sure I can describe the obvious that navigation applications require touch input in the policies even when there's only one of many IVIs with touch capabilities.

**Comment by joeygrover** [comment]
3\. So we can do one of two things or both:

1. Confirm all OEM partners have policies enabled in their production systems and they plan to continue to build with policies enabled. Therefore, we should add a requirement that all production Core builds be built with policies enabled. Or ensure that even with the policy flag turned off, Core still sends the `OnPermissionChange` with all RPCs allowed.
2. As referred to in a later point, we can confirm that all OEMs who use video streaming support touch events and have policies enabled so the app can receive the `OnPermissionChange` notification.


4\. Yes, the scenario where the stream is not enforced to be encrypted but the IVI system has policies enabled will cause a potential blank screen or dialog that needs to appear in the case the IVI is waiting for a PTU. I'm not stating that we need to mitigate every scenario through the library but providing the information in the proposal that this is something that might happen and what the OEM could do about it is fine. For example:

> In the scenario that the Core implementation accepts unencrypted video streams but still has policies enabled, the HMI might need to provide the user with a waiting type dialog before a `StartService` is received. Otherwise a potential blank screen will be displayed to the user until the policy table update has completed and the `OnPermisionChange` notification sent to the application.

**Comment by ashwink11** [comment]
I will be Stand in for Kujtim.

Regarding item 4, if I understood correctly the delay you mentioned will be introduced only if the library holds back the Start Service request for Video until OnTouchEvent permissions are received. However, Kujtim in previous comments mentioned, that if there is no security manager, the library does not need to hold back start service requests. Hence, I suppose, there won't be any delay. The library will delay Start Service request only if security manager is used (protected service is used for Video service), the IVI system that uses unprotected service will stay unaffected.

**Comment by joeygrover** [comment]
@ashwink11 that is incorrect. The proposal states that the `StartService` is held back until the `OnTouchEvent` permission is granted. There is no requirement that video services must be encrypted and therefore checking if a security library exists is irrelevant to if the app has been granted the `OnTouchPermission`.

**Comment by theresalech** [comment]
The Steering Committee voted to defer this proposal until the author is able to respond to the comments on the review issue. The author can email sdlc@smartdevicelink.com when he is available and ready to do so.

**Comment by theresalech** [comment]
The author has advised that he is now available to respond to comments on the review issue, so the Steering Committee can vote to bring this back into review when time allows.

**Comment by theresalech** [comment]
The Steering Committee has voted to bring this proposal back into review. The review will take place until our next meeting, on October 20, 2020 (2020-10-20).

**Comment by theresalech** [comment]
This proposal has been deferred in accordance with [this section](https://github.com/smartdevicelink/sdl_evolution/blob/master/process.md#for-consideration) of the SDL Evolution Process document. The proposal will remain deferred until the author has acknowledged (via email to sdlc@smartdevicelink.com) that they have time to respond to comments in the next week.

**Comment by jordynmackool** [comment]
The author has advised that he is now available to respond to comments on the review issue, so the Steering Committee can vote to bring this back into review when time allows.

**Comment by jordynmackool** [comment]
The author has advised that he is no longer available to respond to comments on the review issue. The proposal will remain deferred until the author has acknowledged (via email to sdlc@smartdevicelink.com) that they have time to respond to comments.

**Comment by jordynmackool** [comment]
The author has advised that he is now available to respond to comments on the review issue, so the Steering Committee can vote to bring this back into review when time allows.

**Comment by jordynmackool** [comment]
The Steering Committee has voted to bring this proposal back into review. The review will take place until our next meeting, on February 02, 2021. (2021-02-02).

**Comment by Sohei-Suzuki-Nexty** [comment]
8. Please let me check for clarification of the proposal.
 a. Is the purpose of this proposal to solve the problem that in the case of the "denylist" method, for an app with an unknown AppID, the process proceeds without obtaining the information defined in the PolicyTable before PolicyUpdate, and an app with an unknown AppID runs without following the defined Permission?
    If so, since this proposal is too focused on onTouchEvent, the big problem is that there are cases where processing is omitted due to asynchronous, and I think it's better to mention that in the case of VideoStream, problems related to this onTouchEvent would occur. Sorry for being late but this is a point for Motivation. I would appreciate your consideration.
 b. The proposal says that NAK will be returned by Core, but what exactly will be the behavior if NAK are returned?
 c. As a fundamental solution, it would be good if the order of application start and PolicyUpdate processing is fixed.
    However I understand that it was necessary to wait for PolicyUpdate to complete when calling startService, probably because it is asynchronous and cannot be fixed etc.
    Is my understanding correct?
 d. Since the focus is on d.onTouchEvent, I feel that the premise of the behavior is unclear. If the app cannot be started for some reason after PolicyUpdate, is there a mechanism to suppress the start by startService etc.?

**Comment by kshala-ford** [comment]
3. So we can do one of two things or both:

 1.   Confirm all OEM partners have policies enabled in their production systems and they plan to continue to build with policies enabled. Therefore, we should add a requirement that all production Core builds be built with policies enabled. Or ensure that even with the policy flag turned off, Core still sends the OnPermissionChange with all RPCs allowed.
   2. As referred to in a later point, we can confirm that all OEMs who use video streaming support touch events and have policies enabled so the app can receive the OnPermissionChange notification.

3\. the suggestions should be treated separately from each other which is not obvious in your response.

3\.1. I don't think we will need to confirm with all OEM partners have policies enabled as Core treats a missing policy manager as a warning, it returns a warning in RAI and it breaks with the policy manager in mobile libraries. The whole SDL ecosystem is not designed to use Core without policies. The question to the SDLC is if the SDL ecosystem should be changed to support what the PM describe as a design oversight. This should be treated separately from this proposal because it clearly doesn't affect this feature and also I expect this change to be very large adding multiple times the effort that this proposal is expected to have. The effort would go beyond what the author and the organization expects. Also, we wouldn't be looking for such an addition as all our IVIs include policy handling. We wouldn't vote against such a change but as we don't see need for our system the change would need to be donated by another entity. If this is not possible I suggest to withdraw this proposal and keep the current status with the design oversight.

3.\2. I also don't think a confirmation is needed. The system already provides touch event capabilities to the app and library allowing the app to behave accordingly.  OEMs can already disallow touch permissions. The only change I see is required is to allow the app specify if requires touch input. The library could also help the user to understand that the system has no touch capabilities but app requires touch input. 

4\. @joeygrover regarding your last comment

> @ashwink11 that is incorrect. The proposal states that the StartService is held back until the OnTouchEvent permission is granted. There is no requirement that video services must be encrypted and therefore checking if a security library exists is irrelevant to if the app has been granted the OnTouchEvent permission.

You are incorrect. Ashwin was not referring to a requirement but, as he mentioned, to my previous comment. See my previous statement:

> The suggestion is to keep the library wait starting the video stream until all three conditions are met. However, if there is no security manager it is not necessary to hold back the service request because it'll likely be accepted. Also it should not be held back if the app doesn't care about touch input.

I highly recommend looking into my previous comment again and provide feedback based on the included suggestions. The proposal revision is meant to allow conversations to suggested changed just like this change.

**Comment by kshala-ford** [comment]
@Sohei-Suzuki-Nexty I am sorry that I overlooked your comment. I'll try to answer your questions:

8.a. The purpose of this proposal is to avoid the attempt to start the video stream as long as not all permissions are given. On systems where the video service is force protected, it is necessary for the IVI and the security manager to have knowledge of the application in the policy database including security assets like TLS certificates for authentication and encryption.

The application won't receive information which services can be started, so it would need to test starting the service and listen for the response. This is an asynchronous task that can add delay to the start routine. Navigation applications are permitted to use OnTouchEvent if the app ID is known so listening for this RPC is a really precise alternative to listen for. On systems where video is force protected the application wouldn't be able to start the service unless OnTouchEvent permission is given. This correlation is what's making this proposal work without adding a delay. I understand there are systems without forced protection and apps that don't need touch input. Alternative paths where suggested for both of these cases.

8.b. The way the libraries are currently implemented, the library tries to start the service once. If it receives a NAK the IVI will remain in the state of no app content. In the case of Ford systems, a popup remains on screen saying "Waiting for device" indefinitely. Other head units might remain just black. You would need to reconnect the phone to reenter the logic again that starts the service.

8.c. Due to the nature of external policies it is not possible to enforce a synchronous process as the policy update must come from an external server. The time until the policy update is received is technically indefinite and cannot be guaranteed to succeed. There are timeouts in place to retry or to return with failure but the process cannot be made synchronous. The best solution is to listen for the asynchronous events and then decide if the video stream can be started. See this comment with an activity diagram showing how to listen for the events which applies well to systems with force protected services. https://github.com/smartdevicelink/sdl_evolution/issues/1014#issuecomment-657500376

8.d. The only correlating information we know of in the current system is the presence of `OnTouchEvent` in the `OnPermissionsChange` RPC. New behavior can be added to the RPC spec or Protocol spec, however it wouldn't solve the problem on existing vehicles.

**Comment by joeygrover** [comment]
> 3\. the suggestions should be treated separately from each other which is not obvious in your response.


   >> 3\. So we can do one of two things or both:
   
As I stated, we can do one of the actions I proposed but we were not limited to doing only one. Either one would solve the issue of incorrectly waiting for the `OnPermissionChange`. I apologize if that wasn't clear for you.

3.1 If we confirm all OEMs use policies then we know the `OnPermissionChange` RPC is sent. The other option with this item is to ensure that Core still sends the OnPermissionChange RPC even if policies are not enabled. 

Regardless of intentions, the execution of how SDL was dumped into the open didn't support policies well with next to no documentation or resources for other members, so I don't believe it is reasonable to assume that the ecosystem is built around something like that. Moving forward I believe we could require it as the PM has created documentation, a sample policy server,  etc. so that members have somewhere to start. However, for this proposal it still must be a consideration that members might not have policies enabled.

The obvious option here is to just check and make sure that Core sends the RPC if policies are disabled, because I know that some OEMs in production do not use policies.

3.2 This is specific to OEMs requiring touch capabilities for video streaming apps or not while also having policies enabled. If not all OEMs require it, then there should be additional changes that allow the app to state if they require touch events or not in the app library API which would circumvent this feature. I find this scenario HIGHLY unlikely, but would be good to make sure we do our due diligence with the SDLC.


4\. The title of this proposal is `Start Video Stream With OnTouchEvent Permission`, which would included both encrypted and unencrypted video streams. Does this proposal intend to fix the scenario where the video stream starts unencrypted and touch events are not enabled or is it not? What you provided states it is not. Therefore I think this proposal should either address that scenario by adding the simple comment I posted or be refactored to the exact scenario which you are requesting, preventing encrypted video streams from starting until the library believes a policy update has happened. I honestly am fine with either, but it should be clear.

   
   
At this point it has been around 5-6 months since I've made my last comments and it is difficult to really follow this review as it has spanned months and months over many comments. I am not sure if my comments here are even on the same path as what I was thinking when initially going through this review. I think it's best to try to summarize where we are at and hopefully return for some agreed upon revisions and restart the review fresh.

**Comment by jordynmackool** [comment]
To provide clarity of the current status of this proposal review, please see the summary below. 

As noted in the last comment from Joey, it is Livio's recommendation that the Steering Committee votes to return this proposal for revisions. Many months have passed since this proposal was first in review and it would be helpful for all those reviewing if the author adds the agreed-upon revisions to the proposal prior to continuing conversation on the remaining open items. 

**Open Items**

3, 4, 8

**Agreed Upon Revisions**

1\.  Add the Java Suite impact to the proposal. 


2\. Remove the code to monitor permissions and add a permissions manager example.  


4b\. Explicitly state that navigation apps require touch input in the policy table.


5\. Core should only send `streamable=true` when both Core and the HMI are ready to accept a StartService for video streaming. This would require a change to the RPC spec that the behavior has changed between when this is implemented and prior. The library can handle such situations, if the RPC version has the refined streamable state, it can use that as its sole determining factor of starting a video stream. However, if the RPC spec version is prior to the changed behavior, the library would need to follow the logic provided in this proposal.


**Resolved (no action required)**

6, 7

**Comment by jordynmackool** [comment]
The Steering Committee voted to return this proposal for revisions. The author is to update the proposal to include the agreed-upon revisions summarized in [this comment](https://github.com/smartdevicelink/sdl_evolution/issues/1014#issuecomment-771979649). When the Steering Committee votes to bring this proposal back "in review" (after the revisions are added) the discussion can continue on the remaining open items (3,4,8).

**Comment by jordynmackool** [comment]
Ford has advised that they are not able to support the effort of this proposal at this time. Should Ford wish to bring this proposal back under Steering Committee review, they can update the pull request to include the agreed-upon revisions and tag Livio requesting a review. Once reviewed/ approved by Livio the Steering Committee can then vote to bring the proposal back into review.  

This issue will be closed for now and remain in a "returned for revisions" state.
